feat(parameters): Configure max_age and decrypt parameters via environment variables

leandrodamascena opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #2041 

## Summary

### Changes

In this PR we changed the utility parameter code to accept environment variables to configure the default TTL and the `decrypt` parameter.

### User experience

Before this PR the customer had to inform the `max_age` parameter every time they wanted a standard TTL different than 5 seconds (default value). If the user had retrieved dozens/hundreds of parameters, this made the developer's experience bad. This PR creates an environment variable called `POWERTOOLS_PARAMETERS_MAX_AGE` that accepts an integer value to be used as the default TTL.

This PR also introduces a new environment variable called `POWERTOOLS_PARAMETERS_SSM_DECRYPT`. This environment variable can be used to set a default value for the `decrypt` parameter in the SSM Provider. This is extremely useful in cases where the customer needs to recover dozens/hundreds of SecureString type parameters and you need to decrypt them. If a value for this environment variable is not informed, the default value `False` remains.

This PR aims to make the developer experience simpler. Below is a code snippet of how a Lambda template can be configured to use these environment variables.

```yaml
Resources:
  HelloWorldFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: HelloWorldFunction
      Handler: app.lambda_handler
      Runtime: python3.9
      Environment:
        Variables:
          POWERTOOLS_PARAMETERS_MAX_AGE: 10
          POWERTOOLS_PARAMETERS_SSM_DECRYPT: true
      Architectures:
      - x86_64
      Events:
        HelloWorld:
          Type: Api
          Properties:
            Path: /hello
            Method: get
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

leandrodamascena committed:
feat(parameters) - environment default variables

leandrodamascena committed:
feat(parameters) - environment default variables - maxage

leandrodamascena commented on aws_lambda_powertools/utilities/parameters/base.py:
```diff
@@ -35,7 +37,9 @@
 
-DEFAULT_MAX_AGE_SECS = 5
+# If the environment variable is not set, the default value is 5
+DEFAULT_MAX_AGE_SECS: int = int(os.getenv(constants.PARAMETERS_MAX_AGE, 5))
+
```
I'm note sure if we should create a function to check if it is integer or not. Currently, we do not control whether `max_age` is an integer anywhere in the code.

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -547,6 +547,44 @@ def test_ssm_provider_get_with_custom_client(mock_name, mock_value, mock_version
         stubber.deactivate()
 
 
+def test_ssm_provider_get_with_decrypt_environment_variable(monkeypatch, mock_name, mock_value, mock_version, config):
```
missing test for `max_age` env var?

leandrodamascena replied:
Hei @heitorlessa, yes! It was missing and refactored the code adding a specific function to resolve the max_age and testing it. I tried to use the function `resolve_env_var_choice` but looks like that mypy doesn't like overload with very similar types like `float` and `int`.

leandrodamascena committed:
feat(parameters) - environment default variables - refactor maxage code

leandrodamascena commented:
@rubenfonseca can you review this PR, please?!

leandrodamascena committed:
Merge branch 'develop' into feat/parameter-env-variable

leandrodamascena committed:
Merge branch 'develop' into feat/parameter-env-variable

rubenfonseca commented on aws_lambda_powertools/shared/constants.py:
```diff
@@ -22,6 +22,9 @@
 IDEMPOTENCY_DISABLED_ENV: str = "POWERTOOLS_IDEMPOTENCY_DISABLED"
 
+PARAMETERS_DEFAULT_DECRYPT: str = "POWERTOOLS_PARAMETERS_SSM_DECRYPT"
+PARAMETERS_MAX_AGE: str = "POWERTOOLS_PARAMETERS_MAX_AGE"
```
I think we should follow the same pattern in the file to have the `_ENV` suffix for constants that represent environment variables:

```suggestion
PARAMETERS_SSM_DECRYPT_ENV: str = "POWERTOOLS_PARAMETERS_SSM_DECRYPT"
PARAMETERS_MAX_AGE_ENV: str = "POWERTOOLS_PARAMETERS_MAX_AGE"
```

Hope you can use your IDE to find/replace this everywhere :D

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/appconfig.py:
```diff
@@ -187,6 +191,8 @@ def get_app_config(
         >>> print(value)
         My configuration's JSON value
     """
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
    # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/base.py:
```diff
@@ -35,7 +38,9 @@
 
 
-DEFAULT_MAX_AGE_SECS = 5
+# If the environment variable is not set, the default value is 5
```
I would remove this comment, as it's confusing as to which environment variable you are refering to.
```suggestion
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/base.py:
```diff
@@ -121,6 +126,9 @@ def get(
         value: Optional[Union[str, bytes, dict]] = None
         key = (name, transform)
 
+        # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
        # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/base.py:
```diff
@@ -186,6 +194,9 @@ def get_multiple(
         """
         key = (path, transform)
 
+        # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
        # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/secrets.py:
```diff
@@ -162,6 +162,9 @@ def get_secret(
         >>> get_secret("my-secret", VersionId="f658cac0-98a5-41d9-b993-8a76a7799194")
     """
 
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
    # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -145,6 +151,14 @@ def get(  # type: ignore[override]
             When the parameter provider fails to transform a parameter value.
         """
 
+        # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
        # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -145,6 +151,14 @@ def get(  # type: ignore[override]
+        # Resolving if will use the default value (5), the value passed by parameter or the environment variable
+        max_age = resolve_max_age(env=os.getenv(constants.PARAMETERS_MAX_AGE, DEFAULT_MAX_AGE_SECS), choice=max_age)
+
+        # Resolving if will use the default value (False), the value passed by parameter or the environment variable
```
```suggestion
        # If decrypt is not set, resolve it from the environment variable, defaulting to false
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -259,6 +273,15 @@ def get_parameters_by_name(
             When "_errors" reserved key is in parameters to be fetched from SSM.
         """
+
+        # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
        # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -259,6 +273,15 @@ def get_parameters_by_name(
+        # Resolving if will use the default value (5), the value passed by parameter or the environment variable
+        max_age = resolve_max_age(env=os.getenv(constants.PARAMETERS_MAX_AGE, DEFAULT_MAX_AGE_SECS), choice=max_age)
+
+        # Resolving if will use the default value (False), the value passed by parameter or the environment variable
```
```suggestion
        # If decrypt is not set, resolve it from the environment variable, defaulting to false
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -543,6 +566,14 @@ def get_parameter(
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
+    max_age = resolve_max_age(env=os.getenv(constants.PARAMETERS_MAX_AGE, DEFAULT_MAX_AGE_SECS), choice=max_age)
+
+    # Resolving if will use the default value (False), the value passed by parameter or the environment variable
```
```suggestion
    # If decrypt is not set, resolve it from the environment variable, defaulting to false
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -543,6 +566,14 @@ def get_parameter(
     if "ssm" not in DEFAULT_PROVIDERS:
         DEFAULT_PROVIDERS["ssm"] = SSMProvider()
 
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
    # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -617,6 +648,14 @@ def get_parameters(
     if "ssm" not in DEFAULT_PROVIDERS:
         DEFAULT_PROVIDERS["ssm"] = SSMProvider()
 
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
    # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -617,6 +648,14 @@ def get_parameters(
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
+    max_age = resolve_max_age(env=os.getenv(constants.PARAMETERS_MAX_AGE, DEFAULT_MAX_AGE_SECS), choice=max_age)
+
+    # Resolving if will use the default value (False), the value passed by parameter or the environment variable
```
```suggestion
    # If decrypt is not set, resolve it from the environment variable, defaulting to false
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -732,6 +771,14 @@ def get_parameters_by_name(
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
+    max_age = resolve_max_age(env=os.getenv(constants.PARAMETERS_MAX_AGE, DEFAULT_MAX_AGE_SECS), choice=max_age)
+
+    # Resolving if will use the default value (False), the value passed by parameter or the environment variable
```
```suggestion
    # If decrypt is not set, resolve it from the environment variable, defaulting to false
```

rubenfonseca commented on aws_lambda_powertools/utilities/parameters/ssm.py:
```diff
@@ -732,6 +771,14 @@ def get_parameters_by_name(
     # NOTE: Decided against using multi-thread due to single-thread outperforming in 128M and 1G + timeout risk
     # see: https://github.com/awslabs/aws-lambda-powertools-python/issues/1040#issuecomment-1299954613
 
+    # Resolving if will use the default value (5), the value passed by parameter or the environment variable
```
```suggestion
    # If max_age is not set, resolve it from the environment variable, defaulting to DEFAULT_MAX_AGE_SECS
```

rubenfonseca commented on docs/index.md:
```diff
@@ -706,6 +706,8 @@ Core utilities such as Tracing, Logging, Metrics, and Event Handler will be avai
 | **POWERTOOLS_LOGGER_LOG_EVENT**           | Logs incoming event                                                                    | [Logging](./core/logger)                                                            | `false`               |
 | **POWERTOOLS_LOGGER_SAMPLE_RATE**         | Debug log sampling                                                                     | [Logging](./core/logger)                                                            | `0`                   |
 | **POWERTOOLS_LOG_DEDUPLICATION_DISABLED** | Disables log deduplication filter protection to use Pytest Live Log feature            | [Logging](./core/logger)                                                            | `false`               |
+| **POWERTOOLS_PARAMETERS_MAX_AGE**         | Adjust how long values are kept in cache (in seconds)                                  | [Parameters](.//utilities/parameters/#adjusting-cache-ttl)                          | `5`                   |
```
```suggestion
| **POWERTOOLS_PARAMETERS_MAX_AGE**         | Adjust how long values are kept in cache (in seconds)                                  | [Parameters](./utilities/parameters/#adjusting-cache-ttl)                          | `5`                   |
```

rubenfonseca commented on docs/index.md:
```diff
@@ -706,6 +706,8 @@ Core utilities such as Tracing, Logging, Metrics, and Event Handler will be avai
 | **POWERTOOLS_LOGGER_SAMPLE_RATE**         | Debug log sampling                                                                     | [Logging](./core/logger)                                                            | `0`                   |
 | **POWERTOOLS_LOG_DEDUPLICATION_DISABLED** | Disables log deduplication filter protection to use Pytest Live Log feature            | [Logging](./core/logger)                                                            | `false`               |
+| **POWERTOOLS_PARAMETERS_MAX_AGE**         | Adjust how long values are kept in cache (in seconds)                                  | [Parameters](.//utilities/parameters/#adjusting-cache-ttl)                          | `5`                   |
+| **POWERTOOLS_PARAMETERS_SSM_DECRYPT**     | Sets whether to decrypt or not values retrieved from AWS SSM Parameters Store          | [Parameters](.//utilities/parameters/#ssmprovider)                                  | `false`               |
```
```suggestion
| **POWERTOOLS_PARAMETERS_SSM_DECRYPT**     | Sets whether to decrypt or not values retrieved from AWS SSM Parameters Store          | [Parameters](./utilities/parameters/#ssmprovider)                                  | `false`               |
```

rubenfonseca commented on docs/utilities/parameters.md:
```diff
@@ -103,7 +103,7 @@ The following will retrieve the latest version and store it in the cache.
 	`max_age` parameter is also available in underlying provider functions like `get()`, `get_multiple()`, etc.
 
-By default, we cache parameters retrieved in-memory for 5 seconds.
+By default, we cache parameters retrieved in-memory for 5 seconds. If you want to change this default value and set the same TTL for all parameters, you can set the `POWERTOOLS_PARAMETERS_MAX_AGE` environment variable. **This will override the default TTL of 5 seconds but can be overridden by the `maxAge` parameter**.
```
```suggestion
By default, we cache parameters retrieved in-memory for 5 seconds. If you want to change this default value and set the same TTL for all parameters, you can set the `POWERTOOLS_PARAMETERS_MAX_AGE` environment variable. **You can still set `max_age` for individual parameters**.
```

rubenfonseca commented on docs/utilities/parameters.md:
```diff
@@ -179,6 +179,9 @@ The AWS Systems Manager Parameter Store provider supports two additional argumen
 You can create `SecureString` parameters, which are parameters that have a plaintext parameter name and an encrypted parameter value. If you don't use the `decrypt` argument, you will get an encrypted value. Read [here](https://docs.aws.amazon.com/kms/latest/developerguide/services-parameter-store.html) about best practices using KMS to secure your parameters.
 
+???+ tip
+	If you want to always decrypt parameters, you can set the `POWERTOOLS_PARAMETERS_SSM_DECRYPT=true` environment variable. **This will override the default value of `false` but can be overridden by the `decrypt` parameter**.
```
```suggestion
	If you want to always decrypt parameters, you can set the `POWERTOOLS_PARAMETERS_SSM_DECRYPT=true` environment variable. **This will override the default value of `false` but you can still set the `decrypt` option for individual parameters**.
```

rubenfonseca reviewed (changes requested):
Great job! It's a little bit tedious because the same logic is present in multiple places. I left some suggestions for improvement.

leandrodamascena committed:
feat(parameters) - addressing Ruben's feedback

leandrodamascena commented:
> Great job! It's a little bit tedious because the same logic is present in multiple places. I left some suggestions for improvement.

Thank you for this super review, Ruben! I really liked it!
I have the same feeling: it's a bit distracting to use the same logic, but it seems like there's no other option (other than refactoring everything) to check if the customer is using the `max_age`/`decrypt` parameters when getting a specific parameter from a specific provider.

I addressed all feedback and let me know if I need to change something else.

leandrodamascena committed:
Merge branch 'develop' into feat/parameter-env-variable

rubenfonseca commented on docs/index.md:
```diff
@@ -706,8 +706,8 @@ Core utilities such as Tracing, Logging, Metrics, and Event Handler will be avai
 | **POWERTOOLS_LOG_DEDUPLICATION_DISABLED** | Disables log deduplication filter protection to use Pytest Live Log feature            | [Logging](./core/logger)                                                            | `false`               |
-| **POWERTOOLS_PARAMETERS_MAX_AGE**         | Adjust how long values are kept in cache (in seconds)                                  | [Parameters](.//utilities/parameters/#adjusting-cache-ttl)                          | `5`                   |
-| **POWERTOOLS_PARAMETERS_SSM_DECRYPT**     | Sets whether to decrypt or not values retrieved from AWS SSM Parameters Store          | [Parameters](.//utilities/parameters/#ssmprovider)                                  | `false`               |
+| **POWERTOOLS_PARAMETERS_MAX_AGE**         | Adjust how long values are kept in cache (in seconds)                                  | [Parameters](./utilities/parameters/#adjusting-cache-ttl)                           | `5`                   |
+| **POWERTOOLS_PARAMETERS_SSM_DECRYPT**     | Sets whether to decrypt or not values retrieved from AWS SSM Parameters Store          | [Parameters](./utilities/parameters/#ssmprovider)                                   | `false`               |
```
I think the environment variable names need to be updated too

leandrodamascena replied:
No no Ruben, we don't need to. In the documentation is the name of the environment variable that will be configured in Lambda, what we changed was the constant in the constant.py file.
We put the constants to make easy the replication of the same value throughout the project's code.

rubenfonseca replied:
I need more coffee

rubenfonseca reviewed (changes requested):
Just a small thing...

rubenfonseca reviewed (approved):
LGTM!
